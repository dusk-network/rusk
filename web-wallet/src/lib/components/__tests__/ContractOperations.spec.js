import { afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/svelte";

import { ContractOperations } from "..";

describe("ContractOperations", () => {
  const baseProps = {
    items: [
      {
        disabled: false,
        id: "send",
        label: "Send",
        primary: true,
      },
      {
        disabled: false,
        id: "receive",
        label: "Receive",
        primary: false,
      },
      {
        disabled: true,
        id: "stake",
        label: "Stake",
        primary: true,
      },
      {
        disabled: false,
        id: "withdraw-stake",
        label: "Withdraw Stake",
        primary: false,
      },
      {
        disabled: false,
        id: "claim-rewards",
        label: "Claim Rewards",
        primary: false,
      },
    ],
  };
  const baseOptions = {
    props: baseProps,
    target: document.body,
  };

  afterEach(cleanup);

  it("should render the `ContractOperations` component", () => {
    const { container } = render(ContractOperations, baseOptions);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should be able to render the component without items", () => {
    const props = {
      ...baseProps,
      items: [],
    };
    const { container } = render(ContractOperations, { ...baseOptions, props });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should use a default icon if the operation is not on the known list", () => {
    const props = {
      ...baseProps,
      items: [
        {
          disabled: false,
          id: "foo-operation",
          label: "Foo operation",
          primary: true,
        },
      ],
    };
    const { container } = render(ContractOperations, { ...baseOptions, props });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should dispatch a "operationChange" event when a operation button is clicked', () => {
    const handleOperationChange = vi.fn();
    const { component, getByRole } = render(ContractOperations, baseOptions);
    const btnReceive = getByRole("button", { name: "Receive" });

    component.$on("operationChange", handleOperationChange);

    fireEvent.click(btnReceive);

    expect(handleOperationChange).toHaveBeenCalledTimes(1);
    expect(handleOperationChange.mock.lastCall?.[0].detail).toBe("receive");
  });
});
