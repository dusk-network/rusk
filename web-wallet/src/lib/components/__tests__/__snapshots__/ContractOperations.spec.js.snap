// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContractOperations > should be able to render the component without items 1`] = `
<menu
  class="contract-operations svelte-1iyylnh"
>
   
  <li
    class="contract-operations__operation svelte-1iyylnh"
  >
    <a
      aria-disabled="false"
      class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default dusk-icon-button--labeled contract-operations__operation-button"
      href="/some-base-path/dashboard"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-anchor-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
        />
      </svg>
      
       
      <span
        class="dusk-anchor-button__text"
      >
        Back
      </span>
      
      
    </a>
  </li>
</menu>
`;

exports[`ContractOperations > should render the \`ContractOperations\` component 1`] = `
<menu
  class="contract-operations svelte-1iyylnh"
>
  <li
    class="contract-operations__operation svelte-1iyylnh"
  >
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--primary dusk-button--size--default dusk-icon-button--labeled contract-operations__operation-button"
      type="button"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M12,7L17,12H14V16H10V12H7L12,7M19,21H5A2,2 0 0,1 3,19V5A2,2 0 0,1 5,3H19A2,2 0 0,1 21,5V19A2,2 0 0,1 19,21M19,19V5H5V19H19Z"
        />
      </svg>
      
       
      <span
        class="dusk-button__text"
      >
        Send
      </span>
      
    </button>
  </li>
  <li
    class="contract-operations__operation svelte-1iyylnh"
  >
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled contract-operations__operation-button"
      type="button"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M12,17L7,12H10V8H14V12H17L12,17M5,3H19A2,2 0 0,1 21,5V19A2,2 0 0,1 19,21H5A2,2 0 0,1 3,19V5A2,2 0 0,1 5,3M5,5V19H19V5H5Z"
        />
      </svg>
      
       
      <span
        class="dusk-button__text"
      >
        Receive
      </span>
      
    </button>
  </li>
  <li
    class="contract-operations__operation svelte-1iyylnh"
  >
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--primary dusk-button--size--default dusk-icon-button--labeled contract-operations__operation-button"
      disabled=""
      type="button"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 3C7.58 3 4 4.79 4 7V17C4 19.21 7.59 21 12 21S20 19.21 20 17V7C20 4.79 16.42 3 12 3M18 17C18 17.5 15.87 19 12 19S6 17.5 6 17V14.77C7.61 15.55 9.72 16 12 16S16.39 15.55 18 14.77V17M18 12.45C16.7 13.4 14.42 14 12 14C9.58 14 7.3 13.4 6 12.45V9.64C7.47 10.47 9.61 11 12 11C14.39 11 16.53 10.47 18 9.64V12.45M12 9C8.13 9 6 7.5 6 7S8.13 5 12 5C15.87 5 18 6.5 18 7S15.87 9 12 9Z"
        />
      </svg>
      
       
      <span
        class="dusk-button__text"
      >
        Stake
      </span>
      
    </button>
  </li>
  <li
    class="contract-operations__operation svelte-1iyylnh"
  >
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled contract-operations__operation-button"
      type="button"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 13.09V7C20 4.79 16.42 3 12 3S4 4.79 4 7V17C4 19.21 7.59 21 12 21C12.46 21 12.9 21 13.33 20.94C13.12 20.33 13 19.68 13 19L13 18.95C12.68 19 12.35 19 12 19C8.13 19 6 17.5 6 17V14.77C7.61 15.55 9.72 16 12 16C12.65 16 13.27 15.96 13.88 15.89C14.93 14.16 16.83 13 19 13C19.34 13 19.67 13.04 20 13.09M18 12.45C16.7 13.4 14.42 14 12 14S7.3 13.4 6 12.45V9.64C7.47 10.47 9.61 11 12 11S16.53 10.47 18 9.64V12.45M12 9C8.13 9 6 7.5 6 7S8.13 5 12 5 18 6.5 18 7 15.87 9 12 9M22 20L19 23L16 20H18V16H20V20H22Z"
        />
      </svg>
      
       
      <span
        class="dusk-button__text"
      >
        Withdraw Stake
      </span>
      
    </button>
  </li>
  <li
    class="contract-operations__operation svelte-1iyylnh"
  >
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--tertiary dusk-button--size--default dusk-icon-button--labeled contract-operations__operation-button"
      type="button"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M22 10.87L19.26 9.28C19.5 9.07 19.73 8.8 19.9 8.5C20.73 7.07 20.24 5.23 18.8 4.4C17.94 3.9 16.93 3.9 16.08 4.26L16.09 4.25L15.21 4.64L15.1 3.68L15.09 3.69C15 2.78 14.47 1.9 13.61 1.4C12.17 .575 10.34 1.07 9.5 2.5C9.33 2.8 9.22 3.13 9.16 3.45L6.41 1.87C5.45 1.32 4.23 1.64 3.68 2.6L2.18 5.2C1.9 5.68 2.07 6.29 2.55 6.56L4.28 7.56L8.5 10H2V20C2 21.11 2.9 22 4 22H20C21.11 22 22 21.11 22 20V14.87L22.73 13.6C23.28 12.64 22.96 11.42 22 10.87M16.44 6.5C16.71 6 17.33 5.86 17.8 6.13C18.28 6.41 18.45 7 18.17 7.5C17.89 8 17.28 8.14 16.8 7.87C16.33 7.59 16.16 7 16.44 6.5M14.07 8.6L21 12.6L20 14.33L13.07 10.33L14.07 8.6M11 20H4V12H11V20M11.34 9.33L4.41 5.33L5.41 3.6L12.34 7.6L11.34 9.33M11.61 4.87C11.13 4.59 10.97 4 11.24 3.5C11.5 3 12.13 2.86 12.61 3.13C13.09 3.41 13.25 4 12.97 4.5C12.7 5 12.09 5.14 11.61 4.87M13 20V12.6L20 16.64V20H13Z"
        />
      </svg>
      
       
      <span
        class="dusk-button__text"
      >
        Claim Rewards
      </span>
      
    </button>
  </li>
   
  <li
    class="contract-operations__operation svelte-1iyylnh"
  >
    <a
      aria-disabled="false"
      class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default dusk-icon-button--labeled contract-operations__operation-button"
      href="/some-base-path/dashboard"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-anchor-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
        />
      </svg>
      
       
      <span
        class="dusk-anchor-button__text"
      >
        Back
      </span>
      
      
    </a>
  </li>
</menu>
`;

exports[`ContractOperations > should use a default icon if the operation is not on the known list 1`] = `
<menu
  class="contract-operations svelte-1iyylnh"
>
  <li
    class="contract-operations__operation svelte-1iyylnh"
  >
    <button
      class="dusk-button dusk-button--type--button dusk-button--variant--primary dusk-button--size--default dusk-icon-button--labeled contract-operations__operation-button"
      type="button"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M11,18H13V16H11V18M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,6A4,4 0 0,0 8,10H10A2,2 0 0,1 12,8A2,2 0 0,1 14,10C14,12 11,11.75 11,15H13C13,12.75 16,12.5 16,10A4,4 0 0,0 12,6Z"
        />
      </svg>
      
       
      <span
        class="dusk-button__text"
      >
        Foo operation
      </span>
      
    </button>
  </li>
   
  <li
    class="contract-operations__operation svelte-1iyylnh"
  >
    <a
      aria-disabled="false"
      class="dusk-anchor dusk-anchor--on-surface dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--default dusk-icon-button--labeled contract-operations__operation-button"
      href="/some-base-path/dashboard"
    >
      <svg
        class="dusk-icon dusk-icon--size--default dusk-anchor-button__icon"
        role="graphics-symbol"
        viewBox="0 0 24 24"
      >
        <path
          d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
        />
      </svg>
      
       
      <span
        class="dusk-anchor-button__text"
      >
        Back
      </span>
      
      
    </a>
  </li>
</menu>
`;
